### ISSUE TITLE
Field Filter Category - Allow no default value for optional clause

### ISSUE THREAD

**Issue by tobbbles** [metabase/metabase#5541]
Category Field Filters should allow no default option, so they can be used properly in optional clauses

# Reproduce
Example clause:
```sql
[[ AND id IN (SELECT id FROM `dataset.with-id-and-name` WHERE {{name}} LIMIT 1) ]]
```
These field filter settings
![2017-07-21-114031_265x308_scrot](https://user-images.githubusercontent.com/10098811/28460322-6e2efda0-6e09-11e7-8bf8-fbb62f3610fa.png)

This 'optional clause' is always included in the native query, even when the widget has no value set.

# Desired Behaviour
There should be a "No Default" field, which is the default (default no default :thinking:)

EG:
![2017-07-21-114832_206x368_scrot](https://user-images.githubusercontent.com/10098811/28460567-95e95f42-6e0a-11e7-831a-d61a2f19fc97.png) ![2017-07-21-114918_263x75_scrot](https://user-images.githubusercontent.com/10098811/28460589-a5471e16-6e0a-11e7-876b-7f57a2d8ec10.png)

When "No Default" is used on this widget, then the clause should be optional and excluded from the native query that's constructed, similar to other, non-fieldfilter widgets.

**Comment by mnzt** [comment]
Here's some handy debug logs, with some info redacted.

Query:
```
#standardSql
SELECT sum(wins) AS wins
FROM `data.secretSauce*`
WHERE {{dateRange}}
[[ AND id IN (SELECT id FROM `data.sauceMakers` WHERE {{name}} LIMIT 1) ]]
ORDER BY wins DESC
```

Debug:
```
{:query
 "#standardSql\nSELECT sum(wins) AS wins\nFROM `aggregate.data*`\nWHERE CAST(date(ts) AS timestamp) BETWEEN ? AND ?\n AND id IN (SELECT id FROM `aggregate.sauceMakers` WHERE 1 = 1 LIMIT 1) \n\nORDER BY wins DESC",
 :template_tags
 {:name {:id "14762412-b62d-DOOT-2f37-863ddad99b84", :name "name", :display_name "Sauce Maker", :type "dimension", :dimension ["field-id" 13330], :widget_type "category"},
  :dateRange
  {:id "333e696b-e258-1dd9-f8b5-e8e4354b9622", :name "dateRange", :display_name "Daterange", :type "dimension", :dimension ["field-id" 1050], :widget_type "date/relative", :default "yesterday"}},
 :params (#inst "2017-06-01T00:00:00.000000000-00:00" #inst "2017-06-30T00:00:00.000000000-00:00")}
```

The bit to focus on here is the subselect from `[[ ... WHERE {{name}} .. ]], as you can see the parameter is substituted with `WHERE 1 = 1 `, which shouldn't be the case; that whole clause should be emitted.

**Comment by mnzt** [comment]
Looks like it's related to this area of code: https://github.com/metabase/metabase/blob/master/src/metabase/query_processor/middleware/parameters/sql.clj#L165 

@camsaul As you were the author of this bit, wanna weigh into what you think's up? :smile_cat:

**Comment by mnzt** [comment]
Digging into the frontend a little I can see that the `ParameterValueWidget` for the variable's default value for the Category dimension has `value: null`.

Whereas if I checkout the the `ParameterValueWidget` for say, DateRange, there is no `value` field.

This seems to add up to my comment :point_up: addressing `what if value is specified but is 'nil'?`

**Comment by sldorman** [comment]
I believe this may be the cause of my problem.  I had an existing SQL question with an optional Where clause and a Text variable.  I was able to leave the text empty and the query worked fine.  I changed the variable type to Field Filter and widget to Category.  Now the query only works if I select something in the parameter drop down or if I set a default value on the parameter.  The error I get is "Incorrect syntax near '='." (SQL Server).  I'm guessing this is because it is using null if no value is set?

A drop down on variables is very helpful but if the variables are not required, the query should ignore them if they are not set.  

Is there a temporary work around?

**Comment by Ucinorn** [comment]
I'd like to bump this bug, as I'm seeing the same behaviour, and I notice that in both cases the variable is within a subquery: 

`SELECT 
  orig.year as year, 
  orig.quarter as quarter, 
  orig.month as month, 
  orig.sum as current_sum, 
  orig.count as current_count, 
  orig.average as current_avg, 
FROM 
(
	SELECT
		year(event_start_datetime) as year, 
		quarter(event_start_datetime) as quarter, 
		month(event_start_datetime) as month, 
		sum(cost) AS sum, 
		count(cost) AS count, 
		sum(cost) / count(cost) AS average 
	FROM 
		_reports_booked_jumpers 
	WHERE 1 = 1
	  [[AND _reports_booked_jumpers.venue = {{venue}}]]
	  [[AND _reports_booked_jumpers.sales_category = {{sales_category}}]]
	GROUP BY 
		year, 
		quarter, 
		month 
	ORDER BY 
		year DESC, 
		quarter DESC, 
		month DESC
) as orig`

And the query i get in the returned is below, not the use of  AND venue = 1 = 1 instead of just omitting the whole line:

`SELECT 
  orig.year as year, 
  orig.quarter as quarter, 
  orig.month as month, 
  orig.sum as current_sum, 
  orig.count as current_count, 
  orig.average as current_avg, 
  prevyear.sum as prevyear_sum, 
  prevyear.count as prevyear_count, 
  prevyear.average as prevyear_avg
FROM 
(
\tSELECT
\t\tyear(event_start_datetime) as year, 
\t\tquarter(event_start_datetime) as quarter, 
\t\tmonth(event_start_datetime) as month, 
\t\tsum(cost) AS sum, 
\t\tcount(cost) AS count, 
\t\tsum(cost) / count(cost) AS average 
\tFROM 
\t\t_reports_booked_jumpers 
    WHERE 1=1
      AND venue = 1 = 1
\tGROUP BY 
\t\tyear, 
\t\tquarter, 
\t\tmonth 
\tORDER BY 
\t\tyear DESC, 
\t\tquarter DESC, 
\t\tmonth DESC
) as orig

**Comment by lp-lima** [comment]
Community post about the same issue:

https://discourse.metabase.com/t/option-clause-for-field-filter-in-native-sql-parameter/6193

This would allow us dinamically hide or show columns according to the user selection

**Comment by camsaul** [comment]
Fixed by #10947
